MAINT: backports and prep for 1.16.0 "final"

tylerjereddy opened the pull request:
I think the current plan is still to release `1.16.0` "final" in the coming week, unless there are objections. June is pretty crazy for me so getting this backport work rolling now.

Backports included (so far):

1. gh-23146
2. gh-23134
3. gh-23127
4. gh-23114
5. gh-23170

TODO:

- [x] check test suite locally
- [x] check doc build/release notes locally
- [x] update the SciPy `1.16.0` release notes following additional backport activity
- [x] remove the "is not released yet" blurb from the release notes on the optimistic assumption we may be able to do the final release
- [ ] make sure the regular CI passes here
- [x] make sure the full wheel build passes here (just to avoid surprises at actual release time)
- [x] @mdhaber wanted `vectorized_filter` considered for the highlights section (we have a lot of highlights, could consider a bit of shuffling around maybe)
- [x] mention `scipy-stubs` `v1.16.0.0` in relnotes

jorenham committed:
DOC: order level 1 blas functions by dtype

jorenham committed:
DOC: Add missing BLAS Level 2 functions

dankiktenko committed:
DOC: fix `linkcode_resolve` `linespec` anchors (#23127)

j-bowhay committed:
BUG: linalg.lapack: fix incorrectly sized work array for {c,z}syrti

longyuxi committed:
DOC: stats.Mixture: add two examples

longyuxi committed:
fixed import issue

mdhaber committed:
MAINT: stats.Mixture: fix PEP8 and doc rendering issues; adjust example

[docs only]

tylerjereddy committed:
DOC: update 1.16.0 relnotes

* Update the SciPy `1.16.0` release notes
following additional backport activity.

* Remove the note about SciPy `1.16.0` not
being released yet.

[ci skip] [skip ci]

tylerjereddy reviewed (commented):
Some minor blemishes in PR titles that might be adjusted from a quick check with `codespell` this evening (probably best to edit the old titles on GH so the script just scrapes that):

```
doc/source/release/1.16.0-notes.rst:488: symetrical ==> symmetrical
doc/source/release/1.16.0-notes.rst:506: optmize ==> optimize
```

this is pretty minor.. anyway I edited those PR titles just now, so this should get corrected when I rerun the script

rgommers committed:
TST: add all SciPy-specific pytest markers to `scipy/conftest.py`

This is necessary in addition to putting them in `pytest.ini`, because
the ini file doesn't get installed and hence running the test suite
with `pytest --pyargs scipy` will otherwise generate a lot of warnings.

[skip circle]

jorenham commented:
I just released `scipy-stubs` [v1.16.0.0](https://github.com/scipy/scipy-stubs/releases/tag/v1.16.0.0). Perhaps we could briefly mention it in the release notes again?

tylerjereddy committed:
DOC: PR 23164 revisions [wheel build]

* Update the SciPy `1.16.0` release notes following
additional backport activity.

* Mention the latest release of `scipy-stubs` in the
SciPy `1.16.0` release notes, per Joren's request.

* Mention the addition of `vectorized_filter` in the
SciPy `1.16.0` release notes "highlights," per Matt H.'s request.

[wheel build]

jorenham commented on doc/source/release/1.16.0-notes.rst:
```diff
@@ -302,6 +302,8 @@ Build and packaging related changes
 *************
 Other changes
 *************
+- A new accompanying release of ``scipy-stubs`` (``v1.16.0.0``) is
+  available.
```
Perhaps some links might help:

```suggestion
- A new accompanying release of `scipy-stubs
  <https://github.com/scipy/scipy-stubs>`_ (`v1.16.0.0 
  <https://github.com/scipy/scipy-stubs/releases/tag/v1.16.0.0>`_) is available.
```

I wasn't able to figure out to keep the inline code style though 🤔

tylerjereddy commented:
Two CI jobs failed or were running for way too long (many hours). The cp312 MacOS ARM wheel build job and the array API testing job. I've cancelled/restarted just those two. It seems hard to believe any of the current changes could be related, so hopefully just something that goes away after a restart.

tylerjereddy commented:
Ah, there are "real" failures in the array API job, I'll see if I can repro locally because the execution there is very slow at the moment.

Recently, versions of the standard or support for newer versions of array libs that may not be pinned have caused occasional problems on the release branch.

tylerjereddy committed:
MAINT, CI: PR 23164 revisions

* Pin the version of `array-api-strict` used
in CI testing because of gh-23183.

tylerjereddy commented:
Local testing suggests that pinning the version of `array-api-strict` below the release version from two days ago should help, as detailed in the issue cross-linked above.

The wheel builds are passing after the job restart, so we only need to see the regular CI pass now to move forward.

tylerjereddy commented:
The single remaining array API test failure is affecting our `main` branch as well, and in gh-23177 appears to be described as an upstream issue in `jax`.

I think in that case I can get away with saying that it is an upstream matter and shouldn't block the release (especially with array API support still "experimental" on this branch).

We may want to pin `jax` version in our release branch CI if that isn't patched "soon," but I don't see it as worth the effort to delay for that right now.
